[Live-Site] Feat/event join peer api

prerana1821 opened the pull request:
# Issues

Fixes #1164 

# Added API:

- [x] PATCH /events/add/peer - Updates the event object with the reference of peer and creates a peer.

TanishqSingla commented on middlewares/validators/events.js:
```diff
@@ -100,11 +100,29 @@ const endActiveEvent = async (req, res, next) => {
 
+const addPeerToEvent = async (req, res, next) => {
+  const schema = joi.object({
+    name: joi.string().required(),
```
Are we having any text validations for these in the backend?

satyam73 replied:
what do you mean by text validation?
here joi is working as a validator.
Anything I am missing?

TanishqSingla commented on test/unit/models/events.test.js:
```diff
@@ -74,4 +76,117 @@ describe("Events", function () {
       }
     });
   });
+
```
nit: can we have have test coverage for `controllers/events.js` and `models/event.js` to be at least 85%?

satyam73 replied:
https://github.com/Real-Dev-Squad/website-backend/pull/1165#discussion_r1238036054

satyam73 commented on models/events.js:
```diff
@@ -66,8 +68,86 @@ const endActiveEvent = async ({ id, reason, lock }) => {
+const addPeerToEvent = async (peerData) => {
+  try {
+    let docRef;
+    const querySnapshot = await peerModel.where("name", "==", peerData.name).limit(1).get();
```
Why are we checking this with name?
Multiple people can have same name

prerana1821 replied:
I understand we shoudn't be checking with names but from frontend I'm only getting name, eventId, role, joinedAt.

prerana1821 replied:
This is resolved as per the new requirements.

satyam73 reviewed (changes requested):
Also when adding peer if I am giving wrong event id it is returning me internal server error with message:
`{
    "message": "Couldn't add peer to the event. Please try again later"
}`
but when requesting again with right info, I see the previous record which was failed is also added.

![image](https://github.com/Real-Dev-Squad/website-backend/assets/85296770/c7e237e4-4f62-44c2-be67-f0d9a461e022)

prerana1821 commented:
> 

Yes, thank you. I have fixed this. I have used the batch way. If any part of the batch operation fails, the changes are rolled back, providing better data consistency.

prerana1821 commented:
> Also when adding peer if I am giving wrong event id it is returning me internal server error with message: `{ "message": "Couldn't add peer to the event. Please try again later" }` but when requesting again with right info, I see the previous record which was failed is also added.
> 
> ![image](https://user-images.githubusercontent.com/85296770/247585823-c7e237e4-4f62-44c2-be67-f0d9a461e022.png)
> 
> here's firestore collection for peer model
> 
> ![image](https://user-images.githubusercontent.com/85296770/247591054-c3d85cb8-f577-4c0a-8570-7288e52eab3f.png)
> 
> index 0 and 2 are failed requests data being added here

Yes, thank you. I have fixed this. I have used the batch way. If any part of the batch operation fails, the changes are rolled back, providing better data consistency.

prerana1821 commented on routes/events.js:
```diff
@@ -10,5 +10,6 @@ router.post("/join", eventsValidator.joinEvent, events.joinEvent);
 router.get("/:id", eventsValidator.getEventById, events.getEventById);
 router.patch("/", authenticate, eventsValidator.updateEvent, events.updateEvent);
 router.patch("/end", authenticate, eventsValidator.endActiveEvent, events.endActiveEvent);
+router.patch("/add/peer", authenticate, eventsValidator.addPeerToEvent, events.addPeerToEvent);
```
Ohk sure. `events/:eventId/peers` what do you think about this?

satyam73 replied:
works👍

satyam73 commented on middlewares/validators/events.js:
```diff
@@ -100,11 +100,29 @@ const endActiveEvent = async (req, res, next) => {
 
+const addPeerToEvent = async (req, res, next) => {
+  const schema = joi.object({
+    name: joi.string().required(),
```
https://github.com/Real-Dev-Squad/website-backend/pull/1165#discussion_r1238036054

shubham-y commented on controllers/events.js:
```diff
@@ -193,11 +193,38 @@ const endActiveEvent = async (req, res) => {
+    });
+  } catch (error) {
+    logger.error({ error });
+    return res.status(500).json({
```
Please use `res.boom.badImplementation` for consistency

prerana1821 replied:
Yes, fixed it. Thank you

shubham-y commented on test/unit/models/events.test.js:
```diff
@@ -74,4 +76,117 @@ describe("Events", function () {
   });
+
+  describe("addPeerToEvent", function () {
+    it("should create a new peer document if it doesn't exist", async function () {
```
NIT 
```suggestion
    it("Should create a new peer document if it doesn't exist", async function () {
```

prerana1821 replied:
"S" is small everywhere. So, following the same for consistency.

shubham-y commented on test/unit/models/events.test.js:
```diff
@@ -74,4 +76,117 @@ describe("Events", function () {
+    it("should create a new peer document if it doesn't exist", async function () {
+      const docRef = await eventModel.add(eventData);
+
+      // Define sample peer data with a non-existing peer name
```
I think these comments are not needed

prerana1821 replied:
Yes, removed.

shubham-y commented on test/unit/models/events.test.js:
```diff
@@ -74,4 +76,117 @@ describe("Events", function () {
+        expect(data.joinedEvents).to.have.lengthOf(1);
+        expect(data.joinedEvents[0].event_id).to.equal(peerData.eventId);
+        expect(data.joinedEvents[0].role).to.equal(peerData.role);
+      } catch (error) {
+        // Check that the function threw an error
+        expect(error).to.not.exist();
+      }
```
I think this catch block is unwanted as the test will fail if there is an error encountered even without this block

prerana1821 replied:
Yes, fixed it.

shubham-y commented on test/unit/models/events.test.js:
```diff
@@ -74,4 +76,117 @@ describe("Events", function () {
     });
   });
+
+  describe("addPeerToEvent", function () {
```
Is there a reason why the integration test for this API is missing?

prerana1821 replied:
As discussed with @satyam73, there are only DB updates here

satyam73 commented on controllers/events.js:
```diff
@@ -193,11 +195,77 @@ const endActiveEvent = async (req, res) => {
+    logger.error({ error });
+    return res.status(500).json({
+      error: error.code,
+      message: "You can't remove selected Participant from Remove, Please ask Admin or Host for help.",
```
please change the message here too

prerana1821 replied:
Done

prerana1821 commented on controllers/events.js:
```diff
@@ -193,11 +195,77 @@ const endActiveEvent = async (req, res) => {
+    });
+    return res.status(200).json({
+      data,
+      message: `Selected participant is removed from event`,
```
Done

vvaibhavdesai commented on controllers/events.js:
```diff
@@ -193,11 +195,77 @@ const endActiveEvent = async (req, res) => {
+  };
+
+  try {
+    await apiService.post(`/active-rooms/${id}/remove-peers`, payload);
```
taking the routes out in constant would become easier to maintain

prerana1821 replied:
Thank you. It is fixed in this PR: https://github.com/Real-Dev-Squad/website-backend/pull/1387

vvaibhavdesai commented on controllers/events.js:
```diff
@@ -193,11 +195,77 @@ const endActiveEvent = async (req, res) => {
+    await apiService.post(`/active-rooms/${id}/remove-peers`, payload);
+    await eventQuery.kickoutPeer({ eventId: id, peerId: payload.peer_id, reason: req.body.reason });
+    return res.status(200).json({
+      message: `Selected Participant is removed from event.`,
```
move your strings to constant

vvaibhavdesai commented on controllers/events.js:
```diff
@@ -193,11 +195,77 @@ const endActiveEvent = async (req, res) => {
+    logger.error({ error });
+    return res.status(500).json({
+      error: error.code,
+      message: "You can't remove selected Participant from Remove, Please ask Admin or Host for help.",
```
same as above

prerana1821 replied:
Fixed here: https://github.com/Real-Dev-Squad/website-backend/pull/1387

vvaibhavdesai commented on middlewares/validators/events.js:
```diff
@@ -100,11 +100,57 @@ const endActiveEvent = async (req, res, next) => {
+  const validationOptions = { abortEarly: false };
+
+  try {
+    await schema.validateAsync({ peerId, name, id, role, joinedAt }, validationOptions);
```
is there a chance of schema validation to fail?

prerana1821 replied:
Yes, it will. If any of the required fields are missing or not provided or if the provided data types for the fields do not match the expected types the validation will fail.

vvaibhavdesai commented on middlewares/validators/events.js:
```diff
@@ -100,11 +100,57 @@ const endActiveEvent = async (req, res, next) => {
+  const validationOptions = { abortEarly: false };
+
+  try {
+    await schema.validateAsync({ id, peerId, reason }, validationOptions);
```
same as above

prerana1821 replied:
Yes, it will fail as discussed above.

prerana1821 commented on controllers/events.js:
```diff
@@ -1,9 +1,11 @@
+/* eslint-disable no-console */
 /* eslint-disable camelcase */
 const { GET_ALL_EVENTS_LIMIT_MIN, UNWANTED_PROPERTIES_FROM_100MS } = require("../constants/events");
```
Thank you. I'll take care from now on.
